Mathlib/AlgebraicGeometry/EllipticCurve/NormalForms.lean

```diff
@@ -0,0 +1,667 @@
+
+  If the field is of characteristic = 3 and the $b_2$ of the curve ≠ 0 (for an elliptic curve,
+  this is equivalent to j ≠ 0), then there exists a change of variables which will change
+  it into such normal form (see `WeierstrassCurve.toCharThreeNF`
+  and `WeierstrassCurve.toCharThreeNF_spec_of_b₂_ne_zero`).
+
+- `WeierstrassCurve.IsCharThreeNF` is the combination of the above two, that is, asserts that
+  a `WeierstrassCurve` is of form $Y^2 = X^3 + a_2X^2 + a_6$ or $Y^2 = X^3 + a_4X + a_6$.
+  It is the normal form of characteristic = 3.
+
+  If the field is of characteristic = 3, then there exists a change of variables which will change
+  it into such normal form (`WeierstrassCurve.exists_variableChange_isCharThreeNF`).
+  See also `WeierstrassCurve.toCharThreeNF` and
+  `WeierstrassCurve.toCharThreeNF_spec`.
+
+- `WeierstrassCurve.IsCharTwoJZeroNF` is a predicate which asserts that a `WeierstrassCurve` is
+  of form $Y^2 + a_3Y = X^3 + a_4X + a_6$. It is the normal form of characteristic = 2 and j = 0.
+
+  If the ring is of characteristic = 2 and the $a_1$ of the curve = 0 (for an elliptic curve,
+  this is equivalent to j = 0), then there exists a change of variables which will change
+  it into such normal form (see `WeierstrassCurve.toCharTwoJZeroNF`
+  and `WeierstrassCurve.toCharTwoJZeroNF_spec`).
+
+- `WeierstrassCurve.IsCharTwoJNeZeroNF` is a predicate which asserts that a `WeierstrassCurve` is
+  of form $Y^2 + XY = X^3 + a_2X^2 + a_6$. It is the normal form of characteristic = 2 and j ≠ 0.
+
+  If the field is of characteristic = 2 and the $a_1$ of the curve ≠ 0 (for an elliptic curve,
+  this is equivalent to j ≠ 0), then there exists a change of variables which will change
+  it into such normal form (see `WeierstrassCurve.toCharTwoJNeZeroNF`
+  and `WeierstrassCurve.toCharTwoJNeZeroNF_spec`).
+
+- `WeierstrassCurve.IsCharTwoNF` is the combination of the above two, that is, asserts that
+  a `WeierstrassCurve` is of form $Y^2 + XY = X^3 + a_2X^2 + a_6$ or
+  $Y^2 + a_3Y = X^3 + a_4X + a_6$.
+  It is the normal form of characteristic = 2.
+
+  If the field is of characteristic = 2, then there exists a change of variables which will change
+  it into such normal form (`WeierstrassCurve.exists_variableChange_isCharTwoNF`).
+  See also `WeierstrassCurve.toCharTwoNF` and
+  `WeierstrassCurve.toCharTwoNF_spec`.
+
+## References
+
+* [J Silverman, *The Arithmetic of Elliptic Curves*][silverman2009]
+
+## Tags
+
+elliptic curve, weierstrass equation, normal form
+
+-/
+
+variable {R : Type*} [CommRing R] (W : WeierstrassCurve R)
+variable {F : Type*} [Field F] (E : EllipticCurve F)
+
+namespace WeierstrassCurve
+
+/-! ### Normal forms of characteristic ≠ 2 -/
+
+/-- A `WeierstrassCurve` is in normal form of characteristic ≠ 2, if its $a_1, a_3 = 0$.
+In other words it is $Y^2 = X^3 + a_2X^2 + a_4X + a_6$. -/
+@[mk_iff]
+structure IsCharNeTwoNF : Prop where
```

Comment:
> Do you think it's better to call this (and the others) `CharNeTwoNF`, so that `toCharNeTwoNF` makes more sense? I'm aware that the naming convention for `Prop`'s is to start with a verb so feel free to ignore this!

I prefer `Is` version, since `CharNeTwoNF` is a little bit misleading: this name suggests that the data itself in a normal form...

> Another suggestion: if you made this into a `class`, then you can include this into the typeclass system, and you can delete all of the `self.*` in the lemmas below.

I'm not sure about this. Dot notation can't be used for `class` approach, and you need to write long names and `haveI`s...

Reply:
Do you need long names and `haveI`'s? At least in this file, except for `IsCharThreeNF`, I replaced all five `structure`'s with `class`'s, replaced all `(self : W. ...)` with `[W. ...]`, removed all `include self` and `self.`, and it seems to work fine (and a lot shorter!). Maybe you have an application in mind, and might be annoying without the dot notation?

If you don't need dot notation, the more natural thing is to remove the namespaces and replace `IsCharNeTwoNF.b₂` with `isCharNeTwoNF_b₂`, etc but maybe you still prefer the dot notation. I don't know which is better.

Reply:
> Do you need long names and `haveI`'s? At least in this file, except for `IsCharThreeNF`, I replaced all five `structure`'s with `class`'s, replaced all `(self : W. ...)` with `[W. ...]`, removed all `include self` and `self.`, and it seems to work fine (and a lot shorter!). Maybe you have an application in mind that is annoying without the dot notation?

I mean outside this file. For example, if you know that a curve is in `IsShortNF`, either by input arguments (`(H : W.IsShortNF)` vs `[W.IsShortNF]`) or by prove it (`have H : W.IsShortNF := by ...` vs `haveI : W.IsShortNF := by ...`). The `structure` and `class` approach make no much difference. Suppose you want to use its properties. For the first case, you write `rw [H.XX]`. For the second case, you need to write `rw [IsShortNF.XX]` or `rw [XX_of_isShortNF]`, which is longer. And you need to remember what normal form you have used.

Moreover, do you think it's a good idea to make `IsShortNF.XX` as simp lemmas? In this way the above `rw` can simply be `simp [H]`. But I'm afraid that this may confuse simp system for `class` approach if you have proved that the curve are in several normal forms. 

> If you don't need dot notation, the more natural thing is to remove the namespaces and replace `IsCharNeTwoNF.b₂` with `isCharNeTwoNF_b₂`, etc but maybe you still prefer the dot notation. I don't know which is better.

Reply:
I've changed it into typeclass, but I'm not sure `simp` works correctly... sometimes I need to manually specify which theorem to be used since it confuses the order.